src/mondemand_backend_annotations_elasticsearch.erl

```diff
@@ -0,0 +1,240 @@
+%% mondemand_server_backend callbacks
+%%====================================================================
+start_link (Config) ->
+  supervisor:start_link ( { local, ?MODULE }, ?MODULE, [Config]).
+
+process (Event) ->
+  mondemand_backend_worker_pool_sup:process (?POOL, Event).
+
+required_apps () ->
+  [ crypto, public_key, ssl, lhttpc, lwes, mondemand ].
+
+type () ->
+  supervisor.
+
+%%====================================================================
+%% supervisor callbacks
+%%====================================================================
+init ([Config]) ->
+  Number = proplists:get_value (number, Config, erlang:system_info(schedulers)), %Was 16
+
+  { ok,
+    {
+      {one_for_one, 10, 10},
+      [
+        { ?POOL,
+          { mondemand_backend_worker_pool_sup, start_link,
+            [ ?POOL,
+              mondemand_backend_worker,
+              Number,
+              ?MODULE ]
+          },
+          permanent,
+          2000,
+          supervisor,
+          [ ]
+        }
+      ]
+    }
+  }.
+
+%%====================================================================
+%% mondemand_backend_worker callbacks
+%%====================================================================
+create (Config) ->
+  AnnotationLabels = proplists:get_value (annotation_labels, Config),
+  CreateUrl = proplists:get_value (create_url, Config),
+  Timeout = proplists:get_value (timeout, Config),
+  UpdateUrl = proplists:get_value (update_url, Config),
+  {ok, #state { create_url = CreateUrl, update_url = UpdateUrl, timeout = Timeout,
+                annotation_labels = AnnotationLabels
+              }}.
+
+connected (#state { create_url = undefined, update_url = undefined }) -> false;
+connected (_) -> true.
+
+connect (State) ->
+  {ok, State}.
+
+send (State = #state {annotation_labels = AnnotationLabels, create_url = Create_Url,
+                      update_url = Update_Url, timeout = Timeout}, Data) ->
+  Annotation = get_annotation(Data, AnnotationLabels),
+  Url = select_url(Annotation, Create_Url, Update_Url),
+  case catch lhttpc:request
+         (Url, "POST",[], format_annotation(Annotation, AnnotationLabels), Timeout, [{max_connections, 256}])
+  of
+    {ok, {{204,_},_,_}} ->
+      { ok, State };
+    E ->
+      error_logger:error_msg ("elasticsearch responded with ~p",[E]),
+      { error, State }
+  end.
+
+destroy (_) ->
+  ok.
+
+%%======================
+%% Internal functions
+%%======================
+get_annotation ( { _Type, _RecTime, _SenderIp, _SenderPort, _Name, MSG }, Labels ) ->
+  { _MsgName, ID, Timestamp, Description, Title, TagCount, Tags, ContextCount, Context } = MSG,
+  TagStr =
+    case Tags of
+      [] -> <<>>;
+      T when is_list (T), T =/= [] ->
+        iolist_to_binary((mondemand_server_util:join (T,",")))
+    end,
+  [ { timestamp, Timestamp }, { description, Description }, { title, Title },
+    { tags, TagStr }, { id, ID } ].
```

Comment:
Sounds good, I'll update that function. Labels was a leftover from a previous refactoring. Tags as a list doesn't work for elasticsearch as the tag field is a string, if a list were passed in, it would send a json array. I don't know what to do with the contexts, I don't know how they fit into the grafana model of an annotation. I suppose we could update the elasticsearch index to have context fields, but they wouldn't be used in grafana currently.